src/main/java/com/thebois/models/world/structures/AbstractStructure.java

```diff
@@ -29,25 +43,60 @@ public Position getPosition() {
     }
 
     @Override
-    public boolean equals(final Object o) {
-        if (this == o) {
-            return true;
+    public StructureType getType() {
+        return structureType;
+    }
+
+    @Override
+    public Collection<ItemType> getNeededItems() {
+        final Collection<ItemType> neededItems = new ArrayList<>();
+        for (final var entry : allNeededItems.entrySet()) {
+            final int numberOf = entry.getValue() - deliveredItems.numberOf(entry.getKey());
+
+            for (int i = 0; i < numberOf; i++) {
+                neededItems.add(entry.getKey());
+            }
         }
-        if (o == null || getClass() != o.getClass()) {
-            return false;
+        return neededItems;
+    }
+
+    private int calculateSumOfAllNeededItems() {
+        int sum = 0;
+        for (final var entry : allNeededItems.entrySet()) {
+            sum += entry.getValue();
         }
-        final AbstractStructure that = (AbstractStructure) o;
-        return position.equals(that.position) && structureType == that.structureType;
+        return sum;
```

Comment:
Does it become more readable though?

How about:
```
        int sum = 0;
        for (final Integer numberOfItems : allNeededItems.values()) {
            sum += numberOfItems;
        }
        return sum;
```

Reply:
Sure, that's an improvement.

I think it might also be a factor of how used you are to reading stream-expressions.

What you basically do is just read it from left to right and ignore the `.stream()` bit. Then you can read the example above as "return all needed item values, then reduce them all into a single integer by summing them together".

I think it's worth getting used to them, as they can be extremely powerful when working with any form of collection, as you saw with my `getNearbyOfType`-example.

If you still don't want to use them, then you can go for the for-loop, that's fine.

Reply:
They're relatively new, as they were introduced in Java 8, but they've since grown to be used quite widely, so definitely something [worth learning about](https://stackify.com/streams-guide-java-8/)!

(C# has a much cleaner version of the same concept, called [LINQ](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/) - which I definitely recommend if you ever write something in C#!)